enhancements/machine-api/control-plane-set.md

```diff
@@ -0,0 +1,603 @@
+ControlPlaneSet resource.
+
+In the future, once we have identified risks and issues with horizontal scale, and mitigated those, we will remove the
+immutability restriction on the replica field to allow horizontal scaling of the control plane between 3 and 5 replicas
+as per current support policy for control plane sizing.
+
+#### Selecting Machines
+
+The ControlPlaneSet operator will use the selector defined within the CRD
+to find Machines which it should consider to be within the set of control plane Machines it is responsible for managing.
+
+This set should be the complete control plane and there should be a 1:1 mapping of Machines in this set to the control
+plane nodes within the cluster.
+
+If there are any control plane nodes (identified by the node role) which do not have a Machine, the operator will mark
+itself degraded as this is an unexpected state.
+No further action will be taken until the unknown node has either been removed from the cluster or a Machine has been
+manually created for it.
+
+#### Providing high availability/fault tolerance within the control plane
+
+Typically within OpenShift, Machines are created within a MachineSet.
+MachineSets have a single `ProviderSpec` which defines the configuration for the Machines created by the MachineSet.
+The failure domain (sometimes called availability zone) is a part of this provider spec and as such, defines that all
+Machines within the Machineset share a failure domain.
+
+This is undesirable for Control Plane Machines as we wish to have them spread across multiple avavailability zones to
+reduce the likelihood of datacenter level faults degrading the control plane.
+To this end, the failure domains for the control plane will be set on the ControlPlaneSet spec directly.
+
+When creating Machines, the ControlPlaneSet controller will balance the Machines across these failure domains by
+injecting the desired failure domain for the new Machine into the provider spec based on the platform specific failure
+domain field.
+
+#### Ensuring Machines match the desired state
+
+To ensure Machines are up to date with the desired configuration, the ControlPlaneSet controller will leverage a
+similar pattern to the workload controllers.
+It will hash the template and compare the hashed result with the spec of the Machines within the managed set.
+As we expect the failure domain to vary across the managed Machines, this will need to be omitted before the hash can
+be calculated.
+
+Should any Machine not match the desired hash, it will be updated based on the chosen update strategy.
+
+##### The RollingUpdate update strategy
+
+The RollingUpdate strategy will be the default. It mirrors the RollingUpdate strategy familiar to users from
+Deployments. When a change is required to be rolled out, the Machine controller will create a new Control Plane
+Machine, wait for this to become ready, and then remove the old Machine. It will repeat this process until all Machines
+are up to date.
```

Comment:
Why does this even matter? With the other enhancement about etcd machine hooks, wouldn't there are always be some kind of rolling update? Or would the machine api controller create the machine anyway and just the etcd membership would be "rolling"? Maybe reference the etcd machine hook enhancement here.

Reply:
I can add a reference for sure

I think the point I was trying to make is that for example, the instance size would be a rolling update as we create and delete machines to migrate over. You're right though, it isn't massively an issue, you could create 3 new machines and then delete all three old ones and etcd's protection mechanism should handle it. I think in previous discussions though we concluded it would be best to only surge a single instance if we can help it, to reduce the likelihood of etcd having issues with balancing the etcd members across zones (the zonal information is under MAPI control not etcd)